# 根据指令进行微调

In [1]:
from importlib.metadata import version

pkgs = [
    "matplotlib",
    "tiktoken",
    "torch",
    "tqdm",
    "tensorflow"
]
for p in pkgs:
    print(f"{p} version: {version(p)}")

matplotlib version: 3.9.0
tiktoken version: 0.7.0
torch version: 2.3.1
tqdm version: 4.66.4
tensorflow version: 2.16.1


![Alt text](../../../img/LLM/ch06/work_flow_finetune.png)

# 1、指令微调简介

- 在之前，我们看到LLM的预训练包括一个训练过程，在这个过程中，LLM一次学习生成一个单词
- 因此，经过预训练的LLM擅长文本完成，但不擅长遵循指令
- 在本章中，我们教LLM更好地遵循说明

![Alt text](../../../img/LLM/ch06/instruction_finetune.png)

- 本章所涵盖的主题总结如下图所示

![Alt text](../../../img/LLM/ch06/topic_summary.png)

# 2、有监督指令微调数据集准备

In [2]:
import json
import os
import urllib

def download_and_load_file(file_path, url):
    if not os.path.exists(file_path):
        with urllib.request.urlopen(url) as response:
            text_data = response.read().decode("utf-8")
        with open(file_path, "w", encoding="utf-8") as file:
            file.write(text_data)
    else:
        with open(file_path, "r", encoding="utf-8") as file:
            text_data = file.read()
    
    with open(file_path, "r") as file:
        data = json.load(file)
    
    return data


file_path = "instruction-data.json"
url = "https://raw.githubusercontent.com/rasbt/LLMs-from-scratch/main/ch07/01_main-chapter-code/instruction-data.json"

data = download_and_load_file(file_path, url)
print("Number of entries:", len(data))

Number of entries: 1100


- 我们从上面的JSON文件加载的数据列表中的每个项都是以下形式的字典

In [3]:
print("Example entry:\n", data[50])

Example entry:
 {'instruction': 'Identify the correct spelling of the following word.', 'input': 'Ocassion', 'output': "The correct spelling is 'Occasion.'"}


- 请注意，“input”字段可以为空：

In [4]:
print("Another example entry:\n", data[999])

Another example entry:
 {'instruction': "What is an antonym of 'complicated'?", 'input': '', 'output': "An antonym of 'complicated' is 'simple'."}


- 指令微调通常被称为“监督指令微调”，因为它涉及在明确提供输入输出对的数据集上训练模型
- 有不同的方式将条目格式化为LLM的输入；下图展示了用于训练Alpaca和Phi-3的两种示例格式

![Alt text](../../../img/LLM/ch06/train_example.png)

- 在本章中，我们使用Alpaca风格的提示格式，这是用于指令微调的原始提示模板
- 下面，我们格式化将作为LLM的输入传递的输入

In [5]:
def format_input(entry):
    instruction_text = (
        f"Below is an instruction that describes a task. "
        f"Write a response that appropriately completes the request."
        f"\n\n### Instruction:\n{entry['instruction']}"
    )

    input_text = f"\n\n### Input:\n{entry['input']}" if entry["input"] else ""

    return instruction_text + input_text

- 带有输入字段的格式化响应如下所示

In [6]:
model_input = format_input(data[50])
desired_response = f"\n\n### Response:\n{data[50]['output']}"

print(model_input + desired_response)

Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Identify the correct spelling of the following word.

### Input:
Ocassion

### Response:
The correct spelling is 'Occasion.'


- 下面是一个没有输入字段的格式化响应

In [7]:
model_input = format_input(data[999])
desired_response = f"\n\n### Response:\n{data[999]['output']}"

print(model_input + desired_response)

Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
What is an antonym of 'complicated'?

### Response:
An antonym of 'complicated' is 'simple'.


- 最后，在下一节中准备PyTorch数据加载程序之前，我们将数据集划分为训练集、验证集和测试集

In [8]:
train_portion = int(len(data) * 0.85)
test_portion = int(len(data) * 0.1)
val_portion = len(data) - train_portion - test_portion

train_data = data[:train_portion]
test_data = data[train_portion:train_portion+test_portion]
val_data = data[train_portion+test_portion:]

In [9]:
print("Training set length:", len(train_data))
print("Validation set length:", len(val_data))
print("Test set length:", len(test_data))

Training set length: 935
Validation set length: 55
Test set length: 110


# 3、将数据组织成训练批次

![Alt text](../../../img/LLM/ch06/Batching_the_dataset.png)

- 我们分几个步骤处理这个数据集批处理，如下图所示

![Alt text](../../../img/LLM/ch06/step_of_batching_dataset.png)

- 首先，我们实现了一个InstructionDataset类，它预标记数据集中的所有输入，类似于第5章中的SpamDataset

![Alt text](../../../img/LLM/ch06/InstructionDataset.png)

In [10]:
import torch
from torch.utils.data import Dataset


class InstructionDataset(Dataset):
    def __init__(self, data, tokenizer):
        self.data = data
        
        self.encoded_texts = []
        for entry in data:
            instruction_plus_input = format_input(entry)
            response_text = f"\n\n### Response:\n{entry['output']}"
            full_text = instruction_plus_input + response_text
            self.encoded_texts.append(
                tokenizer.encode(full_text)
            )
            
    def __getitem__(self, index):
        return self.encoded_texts[index]

    def __len__(self):
        return len(self.data)

- 与第5章类似，我们希望批量收集多个训练示例，以加速训练；这需要将所有输入填充到相似的长度
- 同样与前一章类似，我们使用<|endoftext|>标记作为填充标记

In [14]:
import tiktoken

tokenizer = tiktoken.get_encoding("gpt2")

print(tokenizer.encode("<|endoftext|>", allowed_special={"<|endoftext|>"}))

[50256]


- 在第6章中，我们将数据集中的所有示例填充到相同的长度
    - 在这里，我们采用了一种更复杂的方法，并开发了一个自定义的“collate”函数，我们可以将其传递给数据加载器
    - 此自定义整理功能将每个批次中的训练示例填充为具有相同的长度（但不同批次可以具有不同的长度）

![Alt text](../../../img/LLM/ch06/collate_function.png)

In [15]:
def custome_collate_draft_1(batch, pad_token_id=50256, device="cpu"):
    batch_max_length = max(len(item) + 1 for item in batch)
    
    inputs_lst = []
    
    for item in batch:
        new_item = item.copy()
        new_item += [pad_token_id]
        
        padded = new_item + [pad_token_id] * (batch_max_length - len(new_item))
        
        inputs = torch.tensor(padded[:-1])
        inputs_lst.append(inputs)
    
    inputs_tensor = torch.stack(inputs_lst).to(device)
    return inputs_tensor

In [16]:
inputs_1 = [0, 1, 2, 3, 4]
inputs_2 = [5, 6]
inputs_3 = [7, 8, 9]

batch = (
    inputs_1,
    inputs_2,
    inputs_3
)

print(custome_collate_draft_1(batch))

tensor([[    0,     1,     2,     3,     4],
        [    5,     6, 50256, 50256, 50256],
        [    7,     8,     9, 50256, 50256]])


![Alt text](../../../img/LLM/ch06/token_id_for_training.png)

- 上面，我们只返回LLM的输入；然而，对于LLM训练，我们也需要目标值
- 与预训练LLM类似，目标是向右移动1个位置的输入，因此LLM学习预测下一个token

![Alt text](../../../img/LLM/ch06/predict_token_id.png)

In [17]:
def custome_collate_draft_2(batch, pad_token_id=50256, device="cpu"):
    batch_max_length = max(len(item) + 1 for item in batch)
    
    inputs_lst, targets_lst = [], []
    
    for item in batch:
        new_item = item.copy()
        new_item += [pad_token_id]
        padded = new_item + [pad_token_id] * (batch_max_length - len(new_item))
        inputs = torch.tensor(padded[:-1])
        targets = torch.tensor(padded[1:])
        inputs_lst.append(inputs)
        targets_lst.append(targets)
    
    inputs_tensor = torch.stack(inputs_lst).to(device)
    targets_tensor = torch.stack(targets_lst).to(device)
    return inputs_tensor, targets_tensor

In [18]:
inputs, targets = custome_collate_draft_2(batch)
print(inputs)
print(targets)

tensor([[    0,     1,     2,     3,     4],
        [    5,     6, 50256, 50256, 50256],
        [    7,     8,     9, 50256, 50256]])
tensor([[    1,     2,     3,     4, 50256],
        [    6, 50256, 50256, 50256, 50256],
        [    8,     9, 50256, 50256, 50256]])


- 接下来, 我们引入一个ignore_index值来用一个新值替换所有填充token ID;这个 ignore_index 的目的是我们可以忽略loss函数中的填充值(稍后会详细介绍)

![Alt text](../../../img/LLM/ch06/replace_padding_tokens.png)

- 具体来说，这意味着我们将50256对应的tokenID替换为-100，如下所示

![Alt text](../../../img/LLM/ch06/replace_token_example.png)

- （此外，如果我们想限制样本的长度，我们还引入了allowed_max_length；如果您计划使用自己的数据集，这些数据集的长度超过GPT-2模型支持的1024token上下文大小，这将非常有用。）

In [19]:
def custom_collate_fn(
    batch,
    pad_token_id=50256,
    ignore_index=-100,
    allowed_max_length=None,
    device="cpu"
):
    # Find the longest sequence in the batch
    batch_max_length = max(len(item)+1 for item in batch)

    # Pad and prepare inputs and targets
    inputs_lst, targets_lst = [], []

    for item in batch:
        new_item = item.copy()
        # Add an <|endoftext|> token
        new_item += [pad_token_id]
        # Pad sequences to max_length
        padded = new_item + [pad_token_id] * (batch_max_length - len(new_item))
        inputs = torch.tensor(padded[:-1])  # Truncate the last token for inputs
        targets = torch.tensor(padded[1:])  # Shift +1 to the right for targets

        # New: Replace all but the first padding tokens in targets by ignore_index
        mask = targets == pad_token_id
        indices = torch.nonzero(mask).squeeze()
        if indices.numel() > 1:
            targets[indices[1:]] = ignore_index

        # New: Optionally truncate to maximum sequence length
        if allowed_max_length is not None:
            inputs = inputs[:allowed_max_length]
            targets = targets[:allowed_max_length]

        inputs_lst.append(inputs)
        targets_lst.append(targets)

    # Convert list of inputs and targets to tensors and transfer to target device
    inputs_tensor = torch.stack(inputs_lst).to(device)
    targets_tensor = torch.stack(targets_lst).to(device)

    return inputs_tensor, targets_tensor

In [20]:
inputs, targets = custom_collate_fn(batch)
print(inputs)
print(targets)

tensor([[    0,     1,     2,     3,     4],
        [    5,     6, 50256, 50256, 50256],
        [    7,     8,     9, 50256, 50256]])
tensor([[    1,     2,     3,     4, 50256],
        [    6, 50256,  -100,  -100,  -100],
        [    8,     9, 50256,  -100,  -100]])


- 让我们看看-100的替代能实现什么
- 为了便于说明，让我们假设我们有一个小的分类任务，有两个类标签，0和1，类似于第5章
- 如果我们有以下logits值（模型最后一层的输出），我们计算以下损失

In [21]:
import torch.nn.functional as F
logits_1 = torch.tensor(
    [[-1.0, 1.0],
     [-0.5, 1.5]]
)
targets_1 = torch.tensor([0, 1])

loss_1 = F.cross_entropy(logits_1, targets_1)
print(loss_1)

tensor(1.1269)


- 现在，正如预期的那样，再增加一个训练示例将影响损失

In [22]:
logits_2 = torch.tensor(
    [[-1.0, 1.0],
     [-0.5, 1.5],
     [-0.5, 1.5]]  # New 3rd training example
)
targets_2 = torch.tensor([0, 1, 1])

loss_2 = torch.nn.functional.cross_entropy(logits_2, targets_2)
print(loss_2)

tensor(0.7936)


- 让我们看看如果我们将其中一个示例的类标签替换为-100会发生什么

In [23]:
targets_3 = torch.tensor([0, 1, -100])

loss_3 = torch.nn.functional.cross_entropy(logits_2, targets_3)
print(loss_3)
print("loss_1 == loss_3:", loss_1 == loss_3)

tensor(1.1269)
loss_1 == loss_3: tensor(True)


- 正如我们所看到的，这3个训练示例上的损失与我们从2个训练示例中计算的损失相同，这意味着交叉熵损失函数忽略了带有-100标签的训练示例
- 默认情况下，PyTorch具有cross_entropy（…，ignore_index=-100）设置以忽略与标签-100相对应的示例
- 使用这个-100 ignore_index，我们可以忽略用于将训练示例填充为相等长度的批处理中的额外文本末尾（填充）标记
- 然而，我们不想忽略文本结束（填充）标记（50256）的第一个实例，因为它可以帮助在响应完成时向LLM发出信号
- 在实践中，屏蔽与指令相对应的目标令牌ID也是很常见的，如下图所示（这是完成本章后推荐的读者练习）

![Alt text](../../../img/LLM/ch06/token_id_placeholder.png)

# 4、为指令数据集创建数据加载程序

- 在本节中，我们使用InstructionDataset类和customer_collate_fn函数来实例化训练、验证和测试数据加载程序

![Alt text](../../../img/LLM/ch06/create_data_loaders.png)

- 前面的custom_colate_fn函数的另一个附加细节是，我们现在直接将数据移动到目标设备（例如GPU），而不是在主训练循环中进行，这提高了效率，因为当我们将custom_colat_fn用作数据加载器的一部分时，它可以作为后台进程来执行
- 使用Python的functools标准库中的分部函数，我们创建了一个新函数，其中预先填充了原始函数的设备参数

In [24]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

print(f"Device:{device}")

Device:cpu


In [25]:
from functools import partial

custom_collate_fn = partial(custom_collate_fn, device=device, allowed_max_length=1024)

- 接下来，我们实例化与前几章类似的数据加载程序，只是我们现在为批处理过程提供了自己的collate函数

In [26]:
from torch.utils.data import DataLoader

num_workers = 0
batch_size = 8

torch.manual_seed(123)

train_dataset = InstructionDataset(train_data, tokenizer)
train_loader = DataLoader(
    train_dataset,
    batch_size=batch_size,
    collate_fn=custom_collate_fn,
    shuffle=True,
    drop_last=True,
    num_workers=num_workers
)

val_dataset = InstructionDataset(val_data, tokenizer)
val_loader = DataLoader(
    val_dataset,
    batch_size=batch_size,
    collate_fn=custom_collate_fn,
    shuffle=False,
    drop_last=False,
    num_workers=num_workers
)

test_dataset = InstructionDataset(test_data, tokenizer)
test_loader = DataLoader(
    test_dataset,
    batch_size=batch_size,
    collate_fn=custom_collate_fn,
    shuffle=False,
    drop_last=False,
    num_workers=num_workers
)

In [27]:
print("Train loader:")
for inputs, targets in train_loader:
    print(inputs.shape, targets.shape)

Train loader:
torch.Size([8, 61]) torch.Size([8, 61])
torch.Size([8, 76]) torch.Size([8, 76])
torch.Size([8, 73]) torch.Size([8, 73])
torch.Size([8, 68]) torch.Size([8, 68])
torch.Size([8, 65]) torch.Size([8, 65])
torch.Size([8, 72]) torch.Size([8, 72])
torch.Size([8, 80]) torch.Size([8, 80])
torch.Size([8, 67]) torch.Size([8, 67])
torch.Size([8, 62]) torch.Size([8, 62])
torch.Size([8, 75]) torch.Size([8, 75])
torch.Size([8, 62]) torch.Size([8, 62])
torch.Size([8, 68]) torch.Size([8, 68])
torch.Size([8, 67]) torch.Size([8, 67])
torch.Size([8, 77]) torch.Size([8, 77])
torch.Size([8, 69]) torch.Size([8, 69])
torch.Size([8, 79]) torch.Size([8, 79])
torch.Size([8, 71]) torch.Size([8, 71])
torch.Size([8, 66]) torch.Size([8, 66])
torch.Size([8, 83]) torch.Size([8, 83])
torch.Size([8, 68]) torch.Size([8, 68])
torch.Size([8, 80]) torch.Size([8, 80])
torch.Size([8, 71]) torch.Size([8, 71])
torch.Size([8, 69]) torch.Size([8, 69])
torch.Size([8, 65]) torch.Size([8, 65])
torch.Size([8, 68]) torch.

- 根据上面的输出，我们可以看到，正如预期的那样，所有批次的批次大小都为8，但长度不同
- 还让我们通过在输入批中打印第一个训练示例的内容来双重检查输入是否包含与token ID 50256相对应的<|endoftext|>填充token

In [28]:
print(inputs[0])

tensor([21106,   318,   281, 12064,   326,  8477,   257,  4876,    13, 19430,
          257,  2882,   326, 20431, 32543,   262,  2581,    13,   198,   198,
        21017, 46486,    25,   198, 30003,  6525,   262,  6827,  1262,   257,
          985,   576,    13,   198,   198, 21017, 23412,    25,   198,   464,
         5156,   318,   845, 13779,    13,   198,   198, 21017, 18261,    25,
          198,   464,  5156,   318,   355, 13779,   355,   257,  4936,    13,
        50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256, 50256])


In [29]:
print(targets[0])

tensor([  318,   281, 12064,   326,  8477,   257,  4876,    13, 19430,   257,
         2882,   326, 20431, 32543,   262,  2581,    13,   198,   198, 21017,
        46486,    25,   198, 30003,  6525,   262,  6827,  1262,   257,   985,
          576,    13,   198,   198, 21017, 23412,    25,   198,   464,  5156,
          318,   845, 13779,    13,   198,   198, 21017, 18261,    25,   198,
          464,  5156,   318,   355, 13779,   355,   257,  4936,    13, 50256,
         -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100,  -100])


# 5、加载预训练LLM

![Alt text](../../../img/LLM/ch06/loading_pretrained_LLM.png)

- 然而，我们没有加载最小的1.24亿参数模型，而是加载具有3.55亿参数的中等版本，因为1.24亿模型太小，无法通过指令微调实现质量合理的结果

In [30]:
from gpt_download import download_and_load_gpt2
from ch06 import GPTModel, load_weights_into_gpt


BASE_CONFIG = {
    "vocab_size": 50257,     # Vocabulary size
    "context_length": 1024,  # Context length
    "drop_rate": 0.0,        # Dropout rate
    "qkv_bias": True         # Query-key-value bias
}

model_configs = {
    "gpt2-small (124M)": {"embedding_dim": 768, "n_layers": 12, "n_heads": 12},
    "gpt2-medium (355M)": {"embedding_dim": 1024, "n_layers": 24, "n_heads": 16},
    "gpt2-large (774M)": {"embedding_dim": 1280, "n_layers": 36, "n_heads": 20},
    "gpt2-xl (1558M)": {"embedding_dim": 1600, "n_layers": 48, "n_heads": 25},
}

CHOOSE_MODEL = "gpt2-medium (355M)"

BASE_CONFIG.update(model_configs[CHOOSE_MODEL])

model_size = CHOOSE_MODEL.split(" ")[-1].lstrip("(").rstrip(")")
settings, params = download_and_load_gpt2(model_size=model_size, models_dir="gpt2")

model = GPTModel(BASE_CONFIG)
load_weights_into_gpt(model, params)
model.eval()

File already exists and is up-to-date: gpt2\355M\checkpoint
File already exists and is up-to-date: gpt2\355M\encoder.json
File already exists and is up-to-date: gpt2\355M\hparams.json
File already exists and is up-to-date: gpt2\355M\model.ckpt.data-00000-of-00001
File already exists and is up-to-date: gpt2\355M\model.ckpt.index
File already exists and is up-to-date: gpt2\355M\model.ckpt.meta
File already exists and is up-to-date: gpt2\355M\vocab.bpe


GPTModel(
  (token_emb): Embedding(50257, 1024)
  (pos_emb): Embedding(1024, 1024)
  (drop_emb): Dropout(p=0.0, inplace=False)
  (transformer_blocks): Sequential(
    (0): TransformerBlock(
      (attention): MultiHeadAttention(
        (W_q): Linear(in_features=1024, out_features=1024, bias=True)
        (W_k): Linear(in_features=1024, out_features=1024, bias=True)
        (W_v): Linear(in_features=1024, out_features=1024, bias=True)
        (out_proj): Linear(in_features=1024, out_features=1024, bias=True)
        (dropout): Dropout(p=0.0, inplace=False)
      )
      (ffn): FeedForward(
        (layers): Sequential(
          (0): Linear(in_features=1024, out_features=4096, bias=True)
          (1): GELU()
          (2): Linear(in_features=4096, out_features=1024, bias=True)
        )
      )
      (norm1): LayerNorm()
      (norm2): LayerNorm()
      (drop_shortcut): Dropout(p=0.0, inplace=False)
    )
    (1): TransformerBlock(
      (attention): MultiHeadAttention(
        (W_q):

- 在我们下一节开始微调模型之前，让我们看看它是如何执行其中一个验证任务的

In [31]:
torch.manual_seed(123)

input_text = format_input(val_data[0])
print(input_text)

Below is an instruction that describes a task. Write a response that appropriately completes the request.

### Instruction:
Convert the active sentence to passive: 'The chef cooks the meal every day.'


In [32]:
from ch06 import generate, text_to_token_ids, token_ids_to_text

token_ids = generate(
    model=model,
    idx=text_to_token_ids(input_text, tokenizer),
    max_new_tokens=35,
    context_size=BASE_CONFIG["context_length"],
    eos_id=50256
)
generated_text = token_ids_to_text(token_ids, tokenizer)

- 请注意，我们在前几章中使用的generate函数返回组合的输入和输出文本，这在前一节中很方便创建易读的文本
- 为了隔离响应，我们可以从生成的generated_text的开头减去指令的长度

In [33]:
response_text = generated_text[len(input_text):].strip()
print(response_text)

### Response:

The chef cooks the meal every day.

### Instruction:

Convert the active sentence to passive: 'The chef cooks the


- 正如我们所看到的，该模型还不能遵循说明；它创建了一个“响应”部分，但它只是重复原始输入句子和指令

# 6、在指令数据基础上微调LLM

![Alt text](../../../img/LLM/ch06/Instruction_finetuning_LLM.png)

In [34]:
from ch06 import calc_loss_loader, train_model_simple

ImportError: cannot import name 'train_model_simple' from 'ch06' (D:\LLM\LLM\CourseLearning\ch06\ch06.py)

- 在开始训练之前，让我们计算初始训练和验证集的损失（与前几章一样，目标是尽量减少损失）

In [35]:
model.to(device)

torch.manual_seed(123)

with torch.no_grad():
    train_loss = calc_loss_loader(train_loader, model, device, num_batches=5)
    val_loss = calc_loss_loader(val_loader, model, device, num_batches=5)

print("Training loss:", train_loss)
print("Validation loss:", val_loss)

Training loss: 3.825895595550537
Validation loss: 3.7619208812713625


- 请注意，由于我们使用的是更大的模型（3.55亿而不是1.24亿个参数），因此训练比前几章贵一些
- 下面显示了各种设备的运行时供参考（在兼容的GPU设备上运行此笔记本电脑不需要更改代码）

|Model|Device|Runtime for 2 Epochs|
---|:--:|---:
|gpt2-medium (355M)|CPU (M3 MacBook Air)|15.78 minutes|
|gpt2-medium (355M)|GPU (M3 MacBook Air)|10.77 minutes|
|gpt2-medium (355M)|GPU (L4)|1.83 minutes|
|gpt2-medium (355M)|GPU (A100)|0.86 minutes|
|gpt2-small (124M)|CPU (M3 MacBook Air)|5.74 minutes|
|gpt2-small (124M)|GPU (M3 MacBook Air)|3.73 minutes|
|gpt2-small (124M)|GPU (L4)|0.69 minutes|
|gpt2-small (124M)|GPU (A100)|0.39 minutes|

In [ ]:
import time

start_time = time.time()

torch.manual_seed(123)

optimizer = torch.optim.AdamW(model.parameters(), lr=0.00005, weight_decay=0.1)

num_epochs = 2

train_losses, val_losses, tokens_seen = train_model_simple(
    model, train_loader, val_loader, optimizer, device,
    num_epochs=num_epochs, eval_freq=5, eval_iter=5,
    start_context=format_input(val_data[0]), tokenizer=tokenizer
)

end_time = time.time()
execution_time_minutes = (end_time - start_time) / 60
print(f"Training completed on {execution_time_minutes:.2f} minutes.")

- 根据上述输出，我们可以看到，模型训练良好，从训练损失和验证损失值的减少可以看出
- 此外，根据每个历元后打印的响应文本，我们可以看到模型正确地按照指令转换了输入句子“厨师每天做饭”被动地说：“这顿饭每天都是厨师做的。”（我们将在后面的部分中正确格式化和评估回复）
- 最后，让我们来看看训练和验证损失曲线

In [ ]:
from ch06 import plot_losses

epochs_tensor = torch.linspace(0, num_epochs, len(train_losses))
plot_losses(epochs_tensor, tokens_seen, train_losses, val_losses)

- 正如我们所看到的，损失在第一个纪元开始时急剧下降，这意味着模型开始快速学习
- 我们可以看到，在大约1个训练周期时，会出现轻微的过拟合

# 7、提取并保存响应

![Alt text](../../../img/LLM/ch06/extract_qualitative_evaluation.png)

- 在本节中，我们将保存测试集响应，以便在下一节中评分
- 我们还保存了模型的副本以备将来使用
- 但首先，让我们简要看看微调模型生成的响应

In [ ]:
torch.manual_seed(123)

for entry in test_data[:3]:
    input_text = format_input(entry)
    
    token_ids = generated_text(
        model=model,
        idx=text_to_token_ids(input_text, tokenizer).to(device),
        max_new_tokens=256,
        context_size=BASE_CONFIG["context_length"],
        eos_id=50256
    )
    generated_text = token_ids_to_text(token_ids, tokenizer)
    response_text = (
        generated_text[len(input_text):]
        .replace("### Response:", "")
        .strip()
    )
    
    print(input_text)
    print(f"\nCorrect response:\n>> {entry['output']}")
    print(f"\nModel response:\n>> {response_text.strip()}")
    print("-------------------------------------")

- 正如我们根据测试集指令、给定响应和模型的响应所看到的，模型的性能相对较好

- 第一条和最后一条指令的答案显然是正确的

- 第二个答案很接近；该模型用“积云”而不是“积雨云”来回答（但是，请注意，积雨云可以发展成能够产生雷暴的积雨云）

- 最重要的是，我们可以看到，模型评估并不像前一章那样简单，在前一章中，我们只需计算正确的垃圾邮件/非垃圾邮件类标签的百分比即可获得分类准确性

- 在实践中，通过多种方法评估聊天机器人等指令微调的LLM

    - 诸如MMLU（“测量大规模多任务语言理解”，https://arxiv.org/abs/2009.03300)，测试模型的知识
    - 人类偏好与其他LLM（如LMSYS聊天机器人竞技场）的比较(https://arena.lmsys.org)
    - 自动会话基准测试，其中使用GPT-4等另一种LLM来评估响应，如AlpacaEval(https://tatsu-lab.github.io/alpaca_eval/)
- 在下一节中，我们将使用类似于AlpacaEval的方法，并使用另一种LLM来评估我们模型的响应；但是，我们将使用自己的测试集，而不是使用公开可用的基准数据集

- 为此，我们将模型响应添加到test_data字典中，并将其保存为“带response.json的指令数据”文件进行记录，以便在需要时可以在单独的Python会话中加载和分析它

In [ ]:
from tqdm import tqdm

for i, entry in tqdm(enumerate(test_data), total=len(test_data)):
    input_text = format_input(entry)
    
    token_ids = generated_text(
        model=model,
        idx=text_to_token_ids(input_text, tokenizer).to(device),
        max_new_tokens=256,
        context_size=BASE_CONFIG["context_length"],
        eos_id=50256
    )
    generated_text = token_ids_to_text(token_ids, tokenizer)
    response_text = generated_text[len(input_text):].replace("### Response:", "").strip()
    
    test_data[i]["model_response"] = response_text
    
with open("instruction-data.json", "w") as file:
    json.dump(test_data, file, indent=4)

- 让我们仔细检查其中一个条目，看看响应是否已正确添加到test_data字典中

In [ ]:
print(test_data[0])

- 最后，我们还保存了模型，以备将来重用

In [ ]:
import re

file_name = f"{re.sub(r'[ ()]', '', CHOOSE_MODEL)}-sft.pth"
torch.save(model.state_dict(), file_name)
print(f"Model saved as {file_name}")

# 8、评估微调后的LLM

![Alt text](../../../img/LLM/ch06/evaluate_finetuned_LLM.png)

- 在本节中，我们将使用另一个更大的LLM来自动化微调LLM的响应评估
- 特别是，我们使用Meta AI微调的80亿参数Llama 3模型，该模型可以通过ollama在本地运行(https://ollama.com)
- （或者，如果您更喜欢通过OpenAI API使用功能更强大的LLM，如GPT-4，请参阅LLM-instruction-eval-OpenAI.ipnb笔记本）
- Ollama是一个高效运行LLM的应用程序
- 它是llama.cpp的包装(https://github.com/ggerganov/llama.cpp)，它在纯C/C++中实现LLM，以最大限度地提高效率
- 请注意，它是一个使用LLM生成文本（推理）的工具，而不是训练或微调LLM
- 在运行以下代码之前，请访问以下网址安装ollamahttps://ollama.com并按照说明进行操作（例如，单击“下载”按钮并下载适用于您的操作系统的ollama应用程序）
- 对于macOS和Windows用户，请单击您下载的ollama应用程序；如果它提示您安装命令行用法，请说“是”

- Linux用户可以使用ollama网站上提供的安装命令

- 一般来说，在从命令行使用ollama之前，我们必须启动ollama应用程序或在单独的终端中运行ollama-server

![Alt text](../../../img/LLM/ch06/run_ollama.png)

- 当ollama应用程序或ollama服务在不同的终端中运行时，在命令行上执行以下命令以尝试80亿个参数的Llama 3模型（该模型占用4.7 GB的存储空间，将在您第一次执行此命令时自动下载）

# 8B model
ollama run llama3
The output looks like as follows

    $ ollama run llama3
    pulling manifest
    pulling 6a0746a1ec1a... 100% ▕████████████████▏ 4.7 GB
    pulling 4fa551d4f938... 100% ▕████████████████▏  12 KB
    pulling 8ab4849b038c... 100% ▕████████████████▏  254 B
    pulling 577073ffcc6c... 100% ▕████████████████▏  110 B
    pulling 3f8eb4da87fa... 100% ▕████████████████▏  485 B
    verifying sha256 digest
    writing manifest
    removing any unused layers
    success

- 请注意，llama3是指经过指令微调的80亿Llama 3模型

- 将ollama与“llama3”模型（8B参数模型）一起使用需要16GB的RAM；如果您的计算机不支持此功能，您可以尝试较小的型号，例如通过设置model=“phi-3”来尝试3.8B参数的phi-3型号，这只需要8GB的RAM

- 或者，如果您的机器支持，您也可以使用更大的700亿参数的Llama 3模型，方法是用llama3:70b替换llama3

- 下载完成后，您将看到一个命令行提示符，允许您与模型聊天

- 尝试一个类似“骆驼吃什么？”的提示，它应该返回类似于以下内容的输出
    >>> What do llamas eat?
    Llamas are ruminant animals, which means they have a four-chambered
    stomach and eat plants that are high in fiber. In the wild, llamas
    typically feed on:
    1. Grasses: They love to graze on various types of grasses, including tall
    grasses, wheat, oats, and barley.
    
- 您可以使用输入/bye结束此会话
- 以下代码在继续使用ollama评估我们在上一节中生成的测试集响应之前，检查ollama会话是否正常运行

In [ ]:
import psutil

def check_if_running(process_name):
    running = False
    for proc in psutil.process_iter(["name"]):
        if process_name in proc.info["name"]:
            running = True
            break
    return running

ollama_running = check_if_running("ollama")

if not ollama_running:
    raise RuntimeError("Ollama not running. Launch ollama before proceeding.")
print("Ollama running:", check_if_running("ollama"))

In [ ]:
import json
from tqdm import tqdm

file_path = "instruction-data-with-response.json"

with open(file_path, "r") as file:
    test_data = json.load(file)
    
def format_input(entry):
    instruction_text = (
        f"Below is an instruction that describes a task. "
        f"Write a response that appropriately completes the requests."
        f"\n\n### Instruction:\n{entry['instruction']}"
    )
    
    input_text = f"\n\n### Input:\n{entry['input']}" if entry['input'] else ""
    return instruction_text + input_text

- 现在，我们前面使用的ollama运行命令与模型交互的另一种方法是通过Python中的RESTneneneba API，通过以下函数
- 在运行此笔记本中的下一个单元格之前，请确保ollama仍在运行（前面的代码单元格应打印“ollama running:True”）
- 接下来，运行以下代码单元格以查询模型

In [ ]:
import urllib.request

def query_model(
        prompt,
        model="llama3",
        url="http://localhost:11434/api/chat"
):
    data = {
        "model": model,
        "message": [
            {"role": "user", "content": prompt}
        ],
        "options": {
            "seed": 123,
            "temperature": 0,
            "num_ctx": 2048
        }
    }
    
    payload = json.dumps(data).encode("utf-8")
    
    request = urllib.request.Request(
        url,
        data=payload,
        method="POST"
    )
    request.add_header("Content-Type", "application/json")
    
    response_data = ""
    with urllib.request.urlopen(request) as response:
        while True:
            line = response.readline().decode("utf-8")
            if not line:
                break
            response_json = json.load(line)
            response_data += response_json["message"]["content"]
    return response_data

model = "llama3"
result = query_model("What do Llamas eat?", model)
print(result)

- 现在，使用我们上面定义的query_model函数，我们可以评估我们微调模型的响应；让我们在上一节中查看的前3个测试集响应中尝试一下

In [ ]:
for entry in test_data[:3]:
    prompt = (
        f"Given the input `{format_input(entry)}` "
        f"and correct output `{entry['output']}`, "
        f"score the model response `{entry['model_response']}`"
        f" on a scale from 0 to 100, where 100 is the best score. "
    )
    print("\nDataset response:")
    print(">>", entry['output'])
    print("\nModel response:")
    print(">>", entry["model_response"])
    print("\nScore:")
    print(">>", query_model(prompt))
    print("\n-------------------------")

- 正如我们所看到的，Llama 3模型提供了一个合理的评估，如果模型不完全正确，也会给出部分分数，正如我们在“积云”答案中看到的那样
- 请注意，之前的提示返回了非常详细的评估；我们可以调整提示，生成0到100之间的整数响应（其中100是最好的），以计算模型的平均得分
- 在M3 MacBook Air笔记本电脑上，对测试集中的110个条目进行评估大约需要1分钟

In [ ]:
def generate_model_scores(json_data, json_key, model="llama3"):
    scores = []
    for entry in tqdm(json_data, desc="Scoring entries"):
        prompt = (
            f"Given the input `{format_input(entry)}` "
            f"and correct output `{entry['output']}`, "
            f" on a scale from 0 to 100, where 100 is the best score. "
            f"Respond with the integer number only."
        )
        score = query_model(prompt, model)
        try:
            scores.append(int(score))
        except ValueError:
            print(f"Could not convert score: {score}")
            continue
    return scores

scores = generate_model_scores(test_data, "model_response")
print(f"Number of scores: {len(scores)} of {len(test_data)}")
print(f"Average score: {sum(scores) / len(scores):.2f}\n")

- 我们的模型平均得分超过50，可以作为参考点来比较其他模型或尝试其他训练设置，以改进模型。

- 请注意，Ollama在不同操作系统上并非完全确定性（截至本文撰写时），所以你得到的分数可能会与上面显示的数字略有不同。

    - 作为参考，原始的Llama 3 8B基础模型得分为58.51

    - Llama 3 8B指导模型得分为82.65

# 9、结论

- 这标志着本书的最后一章

- 我们涵盖了LLM开发周期的主要步骤：实现LLM架构、预训练LLM和微调LLM

![Alt text](../../../img/LLM/ch06/all_process_of_book.png)

- 在本章所述的指令微调之后，有时会进行的一个可选步骤是偏好微调

- 偏好微调过程对于使模型更好地符合特定用户偏好特别有用；如果你感兴趣，请参见../04_preference-tuning-with-dpo文件夹

- 此GitHub仓库还包含大量额外的奖励材料，供您享用；欲了解更多信息，请参阅本仓库的README页面中的奖励材料部分